drivers: flash: mchp: Introduce G1 Flash Driver

fabin-mchp opened the pull request:
This pull request adds flash driver support for the Microchip NVMCTRL peripheral. It includes the implementation of the flash driver and updates the device tree for the SAM D5/E5 series to include the necessary node and bindings for the Microchip Nonvolatile Memory Controller (NVMCTRL).

fabin-mchp committed:
drivers: flash: microchip: add flash g1 driver

Add flash driver for Microchip NVMCTRL g1

Signed-off-by: Fabin V Martin <Fabinv.Martin@microchip.com>

fabin-mchp committed:
dts: arm: microchip: add flash node and bindings

Add flash node and binding parameters for nvmctrl
flash driver

Signed-off-by: Fabin V Martin <Fabinv.Martin@microchip.com>

fabin-mchp committed:
boards: microchip: Update sam_e54_xpro device tree

Add zephyr,flash-controller to the chosen node and
add partitions for mcuboot support

Signed-off-by: Fabin V Martin <Fabinv.Martin@microchip.com>

nandojve commented on boards/microchip/sam/sam_e54_xpro/sam_e54_xpro.dts:
```diff
@@ -18,6 +18,7 @@
 		zephyr,shell-uart = &sercom2;
 		zephyr,sram = &sram0;
 		zephyr,flash = &flash0;
+		zephyr,flash-controller = &nvmctrl;
```
We still have a bug in this configuration.

The fact that there is no partition defined, I think, Zephyr will accept the whole amount of flash as a valid program. But, we know that storage_partition will override the last part of the flash.

So, I would like you already exercise the possibility to have MCUboot running and define a flash layout for that.

fabin-mchp replied:
Thanks for the comment!

To clarify, so far we have only added basic flash support but have not yet integrated MCUboot support itself.

Are you asking us to add full MCUboot support now, or can that be done incrementally in subsequent PRs?

fabin-mchp replied:
I've updated the configuration to include a proper flash layout compatible with MCUboot, with explicitly defined partitions. I was also able to successfully run the samples/sysbuild/with_mcuboot, and the output matched the expected behavior described in the README.

nandojve reviewed (changes requested):
Hi @fabin-mchp ,

I did not have time to look in the driver code yet.
I'll look on this soon.

fabin-mchp committed:
samples: sysbuild: update list of allowed platforms

Add sam_e54_xpro board to test mcuboot builds.

Signed-off-by: Fabin V Martin <Fabinv.Martin@microchip.com>

fabin-mchp committed:
tests: boot: test_mcuboot: Add sam_e54_xpro to platform allow list

Enable mcuboot test for sam_e54_xpro board.

Signed-off-by: Fabin V Martin <Fabinv.Martin@microchip.com>

fabin-mchp commented on boards/microchip/sam/sam_e54_xpro/sam_e54_xpro.dts:
```diff
@@ -27,14 +29,25 @@
-		storage_partition: partition@fc000 {
+		boot_partition: partition@0 {
+			label = "mcuboot";
+			reg = <0x00000000 0x00008000>;  /* 32 KB bootloader */
```
For the time being, can we proceed with a 32 kB bootloader size? Allocating a larger boot memory reduces the available space for the application. As our requirements evolve, we can revisit and increase the bootloader size in the future if needed.

fabin-mchp commented on boards/microchip/sam/sam_e54_xpro/sam_e54_xpro.yaml:
```diff
@@ -10,6 +10,7 @@ toolchain:
 flash: 1024
 ram: 256
 supported:
+  - flash
```
I have updated the relevant files to support the sam_e54_xpro board and added the mcuboot tag to the supported list. The changes to the sample and test files have been made in separate commits as requested.

fabin-mchp commented on boards/microchip/sam/sam_e54_xpro/sam_e54_xpro.dts:
```diff
@@ -27,14 +29,25 @@
-		 */
-		storage_partition: partition@fc000 {
+		boot_partition: partition@0 {
+			label = "mcuboot";
```
Removed label.

fabin-mchp commented on boards/microchip/sam/sam_e54_xpro/sam_e54_xpro.yaml:
```diff
@@ -9,7 +9,18 @@ toolchain:
+  - name: image-1
+    size: 480
+  - name: storage
+    size: 32
```
I have removed this section, as the corresponding label has also been removed.

NhMchp commented on drivers/flash/flash_mchp_nvmctrl_g1.c:
```diff
@@ -0,0 +1,1340 @@
+		}
+	}
+
+	if (ret == FLASH_MCHP_SUCCESS) {
```
This check can be clubbed inside ln:741

NhMchp commented on drivers/flash/flash_mchp_nvmctrl_g1.c:
```diff
@@ -0,0 +1,1340 @@
+		}
+	}
+
+	if (ret == FLASH_MCHP_SUCCESS) {
```
No need to check this if previous checks failed. Modify the logic

fabin-mchp replied:
Modified the logic

NhMchp commented on drivers/flash/flash_mchp_nvmctrl_g1.c:
```diff
@@ -0,0 +1,1340 @@
+		ret = -EINVAL;
+	}
+
+	if (ret == FLASH_MCHP_SUCCESS) {
```
No need to check this if previous checks failed. Modify the logic

fabin-mchp commented on drivers/flash/flash_mchp_nvmctrl_g1.c:
```diff
@@ -0,0 +1,1340 @@
+		}
+	}
+
+	if (ret == FLASH_MCHP_SUCCESS) {
```
Updated

fabin-mchp commented on drivers/flash/flash_mchp_nvmctrl_g1.c:
```diff
@@ -0,0 +1,1340 @@
+		ret = flash_check_offset_user_range(address);
+	}
+
+	if (ret == FLASH_MCHP_SUCCESS) {
```
Updated

fabin-mchp commented:
> Hi @fabin-mchp ,
> 
> Could you fix Niyas comments and rebase on top of main ? I would like to see if the CI error will disappear.

Hi @nandojve,

I have fixed the comments provided by Niyas and rebased the branch on top of main.
